File: backend/src/control.cpp
#include "control.hpp"

using namespace Ats;
using namespace std;
using namespace Glib;

Control::Control () : context(1),
                      in_socket(context, ZMQ_REP),
                      out_socket(context, ZMQ_PUB) {

    out_log = IOChannel::create_from_fd(2);
    out_log->set_encoding("");

    /* NOTE: almost all cppzmq calls can throw an exception */

    in_socket.bind("ipc:///tmp/ats_qoe_in");
    out_socket.bind("ipc:///tmp/ats_qoe_out");

    auto read_in = [this](Glib::IOCondition c) -> bool {
        recv();
        return true;
    };

    int in_fd  = in_socket.getsockopt<int>(ZMQ_FD);
    in  = IOChannel::create_from_fd(in_fd);

    Glib::signal_io().connect(read_in, in, Glib::IO_IN);
}

void
Control::recv () {
    int ev = in_socket.getsockopt<int>(ZMQ_EVENTS);
    do {
        if (ev & ZMQ_POLLIN) {
            zmq::message_t m;
            try {
                if (in_socket.recv(&m)) {
                    uint8_t* mptr = static_cast<uint8_t*>(m.data());
                    std::vector<std::uint8_t> data(mptr, mptr + m.size());
                    std::string s = received.emit(data);
                    zmq::message_t reply (s.data(), s.length());
                    in_socket.send(reply);
                }
            } catch (const std::exception& e) {
                log(std::string("Exception while receiving a message: ") + e.what());
            }
        }
        else break;
        ev = in_socket.getsockopt<int>(ZMQ_EVENTS);

    } while (ev & ZMQ_POLLIN);
}

void
Control::send (const string& s) {
    zmq::message_t m(s.data(), s.length());
    out_socket.send(m);
}

void
Control::error (const string& s) {
    send(s); // FIXME?
}

void
Control::log (const string& s) {
    string tmp = "Log: " + s;  
    out_log->write(s);
    out_log->write("\n");
    out_log->flush();
}

File: backend/src/control.rs
use std::thread;
use std::sync::{Arc,Mutex};
use std::sync::mpsc::*;
use signals::Msg;
use zmq;

pub struct Control {
   // out_socket: zmq::Socket,
    pub sender:     Sender<Vec<u8>>,
    pub received: Arc<Mutex<Msg<Vec<u8>,Vec<u8>>>>
}

impl Control {
    
    pub fn new () -> Result<Control,String> {
        let ctx            = zmq::Context::new();
        let in_socket      = ctx.socket(zmq::REP).unwrap();
        let out_socket     = ctx.socket(zmq::PUB).unwrap();
        let (sender, receiver): (Sender<Vec<u8>>, Receiver<Vec<u8>>) = channel();
        let received       = Arc::new(Mutex::new(Msg::new()));
        
        in_socket.connect("ipc:///tmp/ats_qoe_in").unwrap();
        out_socket.connect("ipc:///tmp/ats_qoe_out").unwrap();
        let r = received.clone();
        
        thread::Builder::new().name("recv_thread".to_string()).spawn(move || {
            loop {
                let buffer = in_socket.recv_bytes(0).unwrap();
                let reply : Vec<u8> = r.lock().unwrap().emit(buffer).unwrap();
                in_socket.send (&reply, 0).unwrap();
               // reply.map(|buf| in_socket.send (&buf, 0));
            }
        }).unwrap();

        thread::Builder::new().name("send_thread".to_string()).spawn(move || {
            for msg in receiver {
               // println!("Msg: {}", String::from_utf8_lossy(&msg));
                match out_socket.send(&msg,0) {
                    Ok (()) => debug!("Control::send success"),
                    Err (e) => error!("Control::send error: {}", e),
                }
            }
        }).unwrap();

        Ok (Control { sender, received })
    }

   // pub fn send(&self, buf: &Vec<u8>) {
   //     self.out_socket.send(buf, 0);
   // }
    
    pub fn connect<F> (&mut self, f: F)
        where F: Fn(Vec<u8>) -> Vec<u8> + Send + Sync + 'static {
        self.received.lock().unwrap().connect(f).unwrap()
    }
}
